### ISSUE TITLE
Seemingly wrong short formatting for "de" locale

### ISSUE THREAD

**Issue by epilgrim** [elixir-cldr/cldr_numbers#10]
look at the following outputs:
```
iex(19)> Cldr.Number.to_string(123456, MyApp.Cldr, format: :short, locale: "en")
{:ok, "123K"}
iex(20)> Cldr.Number.to_string(123456, MyApp.Cldr, format: :short, locale: "de")
{:ok, "1"}
```
A lot of information is lost here. No way to differenciate 1 from 100000
But, when I went to the unicode page to create a failing test with the correct representation i found:
![image](https://user-images.githubusercontent.com/203349/59523520-d6872180-8ed1-11e9-8684-3bd2d7666d01.png)
the page is http://www.unicode.org/cldr/charts/latest/summary/de.html#4219
which makes me believe that your library is doing the right thing, but the definition for the german language is wrong.
To compare, here you have the english version: http://www.unicode.org/cldr/charts/latest/summary/en.html#4220

Do you have any clue what might be happening here?

Thanks

**Comment by kipcole9** [comment]
Yes, you are correct in your diagnosis (and really, thanks for taking the time to take a look at the details, it's really appreciated).

The data is:

```elixir
iex> Cldr.Config.get_locale("en", %Cldr.Config{locales: :all}).number_formats.latn.decimal_short
[
  [1000, %{one: ["0K", 1], other: ["0K", 1]}],
  [10000, %{one: ["00K", 2], other: ["00K", 2]}],
  [100000, %{one: ["000K", 3], other: ["000K", 3]}],
  [1000000, %{one: ["0M", 1], other: ["0M", 1]}],
  [10000000, %{one: ["00M", 2], other: ["00M", 2]}],
  [100000000, %{one: ["000M", 3], other: ["000M", 3]}],
  [1000000000, %{one: ["0B", 1], other: ["0B", 1]}],
  [10000000000, %{one: ["00B", 2], other: ["00B", 2]}],
  [100000000000, %{one: ["000B", 3], other: ["000B", 3]}],
  [1000000000000, %{one: ["0T", 1], other: ["0T", 1]}],
  [10000000000000, %{one: ["00T", 2], other: ["00T", 2]}],
  [100000000000000, %{one: ["000T", 3], other: ["000T", 3]}]
]
iex> Cldr.Config.get_locale("de", %Cldr.Config{locales: :all}).number_formats.latn.decimal_short
[
  [1000, %{one: ["0", 1], other: ["0", 1]}],
  [10000, %{one: ["0", 1], other: ["0", 1]}],
  [100000, %{one: ["0", 1], other: ["0", 1]}],
  [1000000, %{one: ["0 Mio'.'", 1], other: ["0 Mio'.'", 1]}],
  [10000000, %{one: ["00 Mio'.'", 2], other: ["00 Mio'.'", 2]}],
  [100000000, %{one: ["000 Mio'.'", 3], other: ["000 Mio'.'", 3]}],
  [1000000000, %{one: ["0 Mrd'.'", 1], other: ["0 Mrd'.'", 1]}],
  [10000000000, %{one: ["00 Mrd'.'", 2], other: ["00 Mrd'.'", 2]}],
  [100000000000, %{one: ["000 Mrd'.'", 3], other: ["000 Mrd'.'", 3]}],
  [1000000000000, %{one: ["0 Bio'.'", 1], other: ["0 Bio'.'", 1]}],
  [10000000000000, %{one: ["00 Bio'.'", 2], other: ["00 Bio'.'", 2]}],
  [100000000000000, %{one: ["000 Bio'.'", 3], other: ["000 Bio'.'", 3]}]
]
```
Which means, as you say, a considerable loss of precision.  I don't have enough personal knowledge to argue about the data quality front.  But I have thought that it should be possible to "patch" a locale by providing a `Map.t` that could be merged into a locale.  Perhaps now is the time to do that?

The idea would be to have something like:

```elixir
defmodule MyApp.Cldr do
  use Cldr, locales: ["en", "de"], 
  providers: [Cldr.Number],
  customize_locales: MyCustomizer
end

defmodule MyCustomizer do
   @behaviour Cldr.Customize

   def customise("de", locale_json) do
      Cldr.Map.deep_merge(locale_json, my_de_customizations())
   end

   def customize(_, locale_json) do
     locale_json
   end

   def my_de_customizations do
      %{number_formats:
        %{latn:
          %{decimal_short:
            [
              [1000, %{one: ["0 K", 1], other: ["0 K", 1]}],
              [10000, %{one: ["00 K", 2], other: ["00 K", 2]}],
              [100000, %{one: ["000 K", 3], other: ["000 K", 3]}],
              [1000000, %{one: ["0 Mio'.'", 1], other: ["0 Mio'.'", 1]}],
              [10000000, %{one: ["00 Mio'.'", 2], other: ["00 Mio'.'", 2]}],
              [100000000, %{one: ["000 Mio'.'", 3], other: ["000 Mio'.'", 3]}],
              [1000000000, %{one: ["0 Mrd'.'", 1], other: ["0 Mrd'.'", 1]}],
              [10000000000, %{one: ["00 Mrd'.'", 2], other: ["00 Mrd'.'", 2]}],
              [100000000000, %{one: ["000 Mrd'.'", 3], other: ["000 Mrd'.'", 3]}],
              [1000000000000, %{one: ["0 Bio'.'", 1], other: ["0 Bio'.'", 1]}],
              [10000000000000, %{one: ["00 Bio'.'", 2], other: ["00 Bio'.'", 2]}],
              [100000000000000, %{one: ["000 Bio'.'", 3], other: ["000 Bio'.'", 3]}]
            ]
          }
        }
      }
  end
end
```
Would that be helpful?

**Comment by epilgrim** [comment]
So, I think we have 2 different problems here:
1. My understanding is that the unicode files are wrong. I'll chase how to report this, and see if it can be fixed (I have no idea, never done this before, might take a day or a year 🤷‍♂to get fixed if it is really a bug)
2. The alternative you are proposing makes sense in the short term to patch existing problems in the upstream. I think the interface `def customize(locale, raw_locale_json)` should be enough, as we are doing really low level and very specific patching here. Your example of deep merge is an alternative, but i would probably just want to modify in some other way the 3 entries i consider wrong. Which way the implementer of the customizer does it, it's his/her problem, and with no warranties if it does it wrongly.

**Comment by epilgrim** [comment]
actually, there seems to be several tickets in their jira about this. 
for example, out of https://unicode-org.atlassian.net/browse/CLDR-9902#icft=CLDR-9902

my understanding is that number 123456 would be "123,456", and only after 1 millon it will show the abbreviation.
But i don't see how one can infer that out of the json data

**Comment by epilgrim** [comment]
it might actually be that it's a but in the library at the end. check te section "short numbers" in this page:
http://cldr.unicode.org/translation/number-patterns
it says:
> If it is impossible to abbreviate numbers in that way in your language, then use a single 0 (zero digit) in your result. For example, the Japanese pattern for 1000 uses that. This should be avoided wherever possible, however, in that it may cause bad displays in devices with small screens, such as mobile phones.

so it looks like a single 0 digit has special meaning, and is "do not do anything"
Am I reading this right?

**Comment by kipcole9** [comment]
Wow, I missed that.  And I thought I understood TR35 pretty well by now!  I've looked again and you are right:

> The default pattern for any type that is not supplied is the special value “0”, as in the following. The value “0” must be used when a child locale overrides a parent locale to drop the compact pattern for that type and use the default pattern.

        <pattern type="1" count="one">0</pattern>

> If the value is precisely “0”, either explicit or defaulted, then the normal number format pattern for that sort of object is supplied — either <decimalFormat> or <currencyFormat type="standard"> — with the normal formatting for the locale (such as the grouping separators). However, for the “0” case by default the signficant digits are adjusted for consistency, typically to 2 or 3 digits, and the maximum fractional digits are set to 0 (for both currencies and plain decimal). Thus the output would be $12, not $12.01. APIs may, however, allow these default behaviors to be overridden.

So it's a bug in my code. Time to go into fixit mode.  Customizations probably still relevant too.

**Comment by kipcole9** [comment]
Fixed with [this commit](https://github.com/elixir-cldr/cldr_numbers/commit/15fe1b3cf949e6d03badbe944dd2f0e191447979).  Published to hex as [ex_cldr_numbers version 2.6.3](https://hex.pm/packages/ex_cldr_numbers/2.6.3).

Appreciate the collaboration - and your patience.

**Comment by epilgrim** [comment]
Whoahhh!! once again, Thank you @kipcole9 !!!!!
that link to TR35 is really useful. I had no idea where you were getting the specifications!!
Have a good day!
